File: app/backend/src/services/auditService.js
import { nanoid } from 'nanoid';
import { mutateStore, readStore } from '../store/jsonStore.js';

export async function audit(action, payload = {}) {
  const entry = {
    id: nanoid(),
    action,
    timestamp: new Date().toISOString(),
    modelVersion: payload.modelVersion || null,
    promptVersion: payload.promptVersion || null,
    request: payload.request || null,
    response: payload.response || null
  };
  await mutateStore((state) => {
    state.audits.unshift(entry);
    state.audits = state.audits.slice(0, 500);
  });
  return entry;
}

export async function listAudit() {
  const state = await readStore();
  return state.audits;
}

File: app/backend/src/main/java/com/bms/service/AuditService.java
package com.bms.service;

import com.bms.repository.JsonStoreRepository;
import com.bms.util.Json;
import org.springframework.stereotype.Service;

import java.time.Instant;
import java.util.ArrayList;
import java.util.Map;

@Service
public class AuditService {
  private final JsonStoreRepository repository;

  public AuditService(JsonStoreRepository repository) {
    this.repository = repository;
  }

  public Map<String, Object> audit(String action, Map<String, Object> payload) {
    Map<String, Object> entry = Json.obj(
        "id", Json.id(),
        "action", action,
        "timestamp", Instant.now().toString(),
        "modelVersion", payload.getOrDefault("modelVersion", null),
        "promptVersion", payload.getOrDefault("promptVersion", null),
        "request", payload.getOrDefault("request", null),
        "response", payload.getOrDefault("response", null));
    repository.mutate(state -> {
      var audits = new ArrayList<>(JsonStoreRepository.collection(state, "audits"));
      audits.add(0, entry);
      state.put("audits", audits.size() > 500 ? audits.subList(0, 500) : audits);
      return entry;
    });
    return entry;
  }

  public Object listAudit() {
    return repository.read().get("audits");
  }
}
